File: student/native/AudioCapture.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Threading;
using NAudio.CoreAudioApi;
using NAudio.Wave;

internal sealed class Pcm16Resampler
{
    private const int TargetRate = 16000;
    private const int PacketSamples = 3200;
    private readonly WaveFormat sourceFormat;
    private readonly double ratio;
    private readonly List<float> pending = new List<float>();
    private readonly short[] packet = new short[PacketSamples];
    private int packetOffset;
    private double readPosition;
    private readonly Stream output;

    public Pcm16Resampler(WaveFormat sourceFormat, Stream output)
    {
        this.sourceFormat = Normalize(sourceFormat);
        this.output = output;
        ratio = (double)this.sourceFormat.SampleRate / TargetRate;
        if (this.sourceFormat.Channels < 1 || this.sourceFormat.SampleRate < TargetRate)
            throw new InvalidOperationException("Unsupported audio format: " + this.sourceFormat);
    }

    private static WaveFormat Normalize(WaveFormat format)
    {
        WaveFormatExtensible extensible = format as WaveFormatExtensible;
        return extensible == null ? format : extensible.ToStandardWaveFormat();
    }

    public void Write(byte[] buffer, int count)
    {
        int bytesPerSample = sourceFormat.BitsPerSample / 8;
        int frameBytes = bytesPerSample * sourceFormat.Channels;
        if (bytesPerSample < 2 || frameBytes <= 0) throw new InvalidOperationException("Unsupported audio format: " + sourceFormat);

        for (int offset = 0; offset + frameBytes <= count; offset += frameBytes)
        {
            float mono = 0;
            for (int channel = 0; channel < sourceFormat.Channels; channel++)
                mono += ReadSample(buffer, offset + channel * bytesPerSample, bytesPerSample);
            pending.Add(mono / sourceFormat.Channels);
        }

        while (readPosition + 1 < pending.Count)
        {
            int left = (int)readPosition;
            double fraction = readPosition - left;
            double sample = pending[left] * (1 - fraction) + pending[left + 1] * fraction;
            sample = Math.Max(-1, Math.Min(1, sample));
            packet[packetOffset++] = (short)Math.Round(sample < 0 ? sample * 32768 : sample * 32767);
            readPosition += ratio;
            if (packetOffset == PacketSamples) FlushPacket();
        }

        int consumed = Math.Min((int)readPosition, Math.Max(0, pending.Count - 1));
        if (consumed > 0)
        {
            pending.RemoveRange(0, consumed);
            readPosition -= consumed;
        }
    }

    private float ReadSample(byte[] buffer, int offset, int bytesPerSample)
    {
        if (sourceFormat.Encoding == WaveFormatEncoding.IeeeFloat && bytesPerSample == 4)
            return BitConverter.ToSingle(buffer, offset);
        if (sourceFormat.Encoding != WaveFormatEncoding.Pcm)
            throw new InvalidOperationException("Unsupported audio encoding: " + sourceFormat.Encoding);
        if (bytesPerSample == 2) return BitConverter.ToInt16(buffer, offset) / 32768f;
        if (bytesPerSample == 3)
        {
            int value = buffer[offset] | (buffer[offset + 1] << 8) | (buffer[offset + 2] << 16);
            if ((value & 0x800000) != 0) value |= unchecked((int)0xff000000);
            return value / 8388608f;
        }
        if (bytesPerSample == 4) return BitConverter.ToInt32(buffer, offset) / 2147483648f;
        throw new InvalidOperationException("Unsupported PCM bit depth: " + sourceFormat.BitsPerSample);
    }

    private void FlushPacket()
    {
        byte[] bytes = new byte[PacketSamples * 2];
        Buffer.BlockCopy(packet, 0, bytes, 0, bytes.Length);
        output.Write(bytes, 0, bytes.Length);
        output.Flush();
        packetOffset = 0;
    }
}

internal static class Program
{
    private static WasapiLoopbackCapture capture;
    private static MMDevice device;
    private static readonly ManualResetEvent stopped = new ManualResetEvent(false);
    private static Exception captureError;

    private static string JsonString(string value)
    {
        if (value == null) return "null";
        StringBuilder result = new StringBuilder("\"");
        foreach (char character in value)
        {
            switch (character)
            {
                case '\\': result.Append("\\\\"); break;
                case '"': result.Append("\\\""); break;
                case '\r': result.Append("\\r"); break;
                case '\n': result.Append("\\n"); break;
                case '\t': result.Append("\\t"); break;
                default:
                    if (character < 32) result.Append("\\u" + ((int)character).ToString("x4"));
                    else result.Append(character);
                    break;
            }
        }
        return result.Append('"').ToString();
    }

    private static void Status(string eventName, string message, WaveFormat format)
    {
        string formatFields = format == null ? "" :
            ",\"sampleRate\":" + format.SampleRate + ",\"channels\":" + format.Channels + ",\"bits\":" + format.BitsPerSample;
        Console.Error.WriteLine("{\"event\":" + JsonString(eventName) + ",\"message\":" + JsonString(message) +
            ",\"device\":" + JsonString(device == null ? null : device.FriendlyName) + formatFields + "}");
        Console.Error.Flush();
    }

    public static int Main()
    {
        Console.OutputEncoding = Encoding.UTF8;
        try
        {
            device = WasapiLoopbackCapture.GetDefaultLoopbackCaptureDevice();
            capture = new WasapiLoopbackCapture(device);
            Pcm16Resampler resampler = new Pcm16Resampler(capture.WaveFormat, Console.OpenStandardOutput());
            capture.DataAvailable += delegate(object sender, WaveInEventArgs args)
            {
                try { resampler.Write(args.Buffer, args.BytesRecorded); }
                catch (Exception error) { captureError = error; stopped.Set(); }
            };
            capture.RecordingStopped += delegate(object sender, StoppedEventArgs args)
            {
                if (args.Exception != null) captureError = args.Exception;
                stopped.Set();
            };
            capture.StartRecording();
            Status("ready", "WASAPI loopback capture started", capture.WaveFormat);

            Thread inputThread = new Thread(delegate()
            {
                try { Console.In.ReadLine(); }
                catch { }
                stopped.Set();
            });
            inputThread.IsBackground = true;
            inputThread.Start();
            stopped.WaitOne();
            capture.StopRecording();
            if (captureError != null) throw captureError;
            Status("stopped", "Capture stopped", null);
            return 0;
        }
        catch (Exception error)
        {
            Status("error", error.Message, null);
            return 1;
        }
        finally
        {
            if (capture != null) capture.Dispose();
            if (device != null) device.Dispose();
        }
    }
}

File: student/native/AudioCapture.swift
import AudioToolbox
import CoreMedia
import Foundation
import ScreenCaptureKit

private let targetSampleRate = 16_000.0
private let packetSamples = 3_200

private func writeStatus(_ event: String, message: String, extra: [String: Any] = [:]) {
    var payload: [String: Any] = ["event": event, "message": message]
    extra.forEach { payload[$0.key] = $0.value }
    guard let data = try? JSONSerialization.data(withJSONObject: payload),
          var line = String(data: data, encoding: .utf8) else { return }
    line.append("\n")
    FileHandle.standardError.write(Data(line.utf8))
}

private final class PacketWriter {
    private var pending = Data()

    func append(_ samples: [Int16]) {
        samples.withUnsafeBytes { pending.append(contentsOf: $0) }
        let packetBytes = packetSamples * MemoryLayout<Int16>.size
        while pending.count >= packetBytes {
            FileHandle.standardOutput.write(pending.prefix(packetBytes))
            pending.removeFirst(packetBytes)
        }
    }
}

private final class Pcm16Resampler {
    private let packetWriter: PacketWriter
    private var pending = [Float]()
    private var readPosition = 0.0
    private var sourceSampleRate = 0.0

    init(packetWriter: PacketWriter) {
        self.packetWriter = packetWriter
    }

    func append(_ source: [Float], sampleRate: Double) {
        guard sampleRate > 0 else { return }
        if sourceSampleRate != sampleRate {
            sourceSampleRate = sampleRate
            pending.removeAll(keepingCapacity: true)
            readPosition = 0
        }
        pending.append(contentsOf: source)

        let ratio = sourceSampleRate / targetSampleRate
        var output = [Int16]()
        output.reserveCapacity(Int(Double(source.count) / ratio) + 1)
        while readPosition + 1 < Double(pending.count) {
            let left = Int(readPosition)
            let fraction = Float(readPosition - Double(left))
            let value = max(-1.0, min(1.0, pending[left] * (1 - fraction) + pending[left + 1] * fraction))
            output.append(Int16(value < 0 ? value * 32768.0 : value * 32767.0))
            readPosition += ratio
        }

        let consumed = min(Int(readPosition), max(0, pending.count - 1))
        if consumed > 0 {
            pending.removeFirst(consumed)
            readPosition -= Double(consumed)
        }
        packetWriter.append(output)
    }
}

private final class SystemAudioCapture: NSObject, SCStreamOutput, SCStreamDelegate {
    private let outputQueue = DispatchQueue(label: "cn.internal.interviewassistant.audio")
    private let packetWriter = PacketWriter()
    private lazy var resampler = Pcm16Resampler(packetWriter: packetWriter)
    private var stream: SCStream?
    private var stopped = false

    func start() async throws -> String {
        let content = try await SCShareableContent.excludingDesktopWindows(
            false,
            onScreenWindowsOnly: true
        )
        guard let display = content.displays.first else {
            throw NSError(
                domain: "InterviewAudioCapture",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "No display is available for system audio capture"]
            )
        }

        let filter = SCContentFilter(
            display: display,
            excludingApplications: [],
            exceptingWindows: []
        )
        let configuration = SCStreamConfiguration()
        configuration.capturesAudio = true
        configuration.excludesCurrentProcessAudio = true
        configuration.sampleRate = Int(targetSampleRate)
        configuration.channelCount = 1
        configuration.width = 2
        configuration.height = 2
        configuration.minimumFrameInterval = CMTime(value: 1, timescale: 1)
        configuration.showsCursor = false
        configuration.queueDepth = 3

        let stream = SCStream(filter: filter, configuration: configuration, delegate: self)
        try stream.addStreamOutput(self, type: .audio, sampleHandlerQueue: outputQueue)
        self.stream = stream
        try await stream.startCapture()
        return display.displayID.description
    }

    func stop() async {
        guard !stopped else { return }
        stopped = true
        guard let stream else { return }
        try? await stream.stopCapture()
        self.stream = nil
    }

    func stream(_ stream: SCStream, didStopWithError error: Error) {
        guard !stopped else { return }
        stopped = true
        writeStatus("error", message: error.localizedDescription)
        exit(1)
    }

    func stream(
        _ stream: SCStream,
        didOutputSampleBuffer sampleBuffer: CMSampleBuffer,
        of outputType: SCStreamOutputType
    ) {
        guard outputType == .audio,
              sampleBuffer.isValid,
              sampleBuffer.dataReadiness == .ready,
              let format = sampleBuffer.formatDescription,
              let description = CMAudioFormatDescriptionGetStreamBasicDescription(format)?.pointee else { return }

        guard description.mFormatID == kAudioFormatLinearPCM else {
            writeStatus("error", message: "ScreenCaptureKit returned an unsupported audio format")
            Task { await stop(); exit(1) }
            return
        }

        var bufferList = AudioBufferList()
        var blockBuffer: CMBlockBuffer?
        let status = CMSampleBufferGetAudioBufferListWithRetainedBlockBuffer(
            sampleBuffer,
            bufferListSizeNeededOut: nil,
            bufferListOut: &bufferList,
            bufferListSize: MemoryLayout<AudioBufferList>.size,
            blockBufferAllocator: kCFAllocatorDefault,
            blockBufferMemoryAllocator: kCFAllocatorDefault,
            flags: UInt32(kCMSampleBufferFlag_AudioBufferList_Assure16ByteAlignment),
            blockBufferOut: &blockBuffer
        )
        guard status == noErr, let rawData = bufferList.mBuffers.mData else { return }

        let frameCount = sampleBuffer.numSamples
        let isFloat = description.mFormatFlags & kAudioFormatFlagIsFloat != 0
        let isNonInterleaved = description.mFormatFlags & kAudioFormatFlagIsNonInterleaved != 0
        let bitsPerChannel = Int(description.mBitsPerChannel)
        let channels = Int(description.mChannelsPerFrame)
        guard channels > 0, channels == 1 || !isNonInterleaved else {
            writeStatus("error", message: "ScreenCaptureKit returned unsupported planar multichannel audio")
            Task { await stop(); exit(1) }
            return
        }
        var samples = [Float]()
        samples.reserveCapacity(frameCount)

        if isFloat && bitsPerChannel == 32 {
            let values = rawData.assumingMemoryBound(to: Float.self)
            for index in 0..<frameCount {
                var mono: Float = 0
                for channel in 0..<channels { mono += values[index * channels + channel] }
                samples.append(mono / Float(channels))
            }
        } else if !isFloat && bitsPerChannel == 16 {
            let values = rawData.assumingMemoryBound(to: Int16.self)
            for index in 0..<frameCount {
                var mono: Float = 0
                for channel in 0..<channels {
                    mono += Float(values[index * channels + channel]) / 32768.0
                }
                samples.append(mono / Float(channels))
            }
        } else {
            writeStatus("error", message: "ScreenCaptureKit returned an unsupported PCM representation")
            Task { await stop(); exit(1) }
            return
        }

        resampler.append(samples, sampleRate: description.mSampleRate)
    }
}

@main
private enum InterviewAudioCapture {
    static func main() async {
        let capture = SystemAudioCapture()
        do {
            let display = try await capture.start()
            writeStatus(
                "ready",
                message: "ScreenCaptureKit system audio capture started",
                extra: ["device": "Display \(display)", "sampleRate": 16_000, "channels": 1, "bits": 16]
            )
            await withCheckedContinuation { continuation in
                DispatchQueue.global(qos: .utility).async {
                    _ = readLine()
                    continuation.resume()
                }
            }
            await capture.stop()
            writeStatus("stopped", message: "Capture stopped")
        } catch {
            writeStatus("error", message: error.localizedDescription)
            exit(1)
        }
    }
}
